### ISSUE TITLE
Sparse tensor conversion performance issues (CPU/GPU)

### ISSUE THREAD

**Issue by aartbik** [pytorch/pytorch#151351]
### 🐛 Describe the bug

I was investigating opportunities for improving "activation" sparsity as follows

```
import time
import torch

d = 1024 * 32
A = -torch.ones(d, d)
A[0, 0] = 111
A[10, 10] = 222     # most entries in A are < 0

T = torch.relu(A)         # materializes very sparse T as dense first
S = T.to_sparse_csr()     # would be nice to have something like S = torch.sparse_relu(A)
                          # but that is not the point of this bug yet
```

where I really would like to have a "sparsifying" relu to avoid materializing T as dense intermediate first.

However, while pondering on that, I noticed a huge performance (time and memory) difference between converting to COO or converting to CSR. Take the following annotated code

```
.. construct A..
time.sleep(10)      # INTERVAL 1
T = torch.relu(A)   # materializes very sparse T as dense first
time.sleep(10)      # INTERVAL 2
S = T.to_sparse()   # to COO
time.sleep(10)      # INTERVAL 3
```

and compare memory tracking for COO (to_sparse) with CSR (to_sparse_csr). Attached are the two plots. Both are running on CPU (but the problem occurs for GPU as well, same code path).

It is clear that COO behaves more or less as expected (first bump to get A, second bump to get T, then nothing more for S; again, my initial goal was to avoid the second bump for T, but read on). Then looking at CSR, we get a similar first bump to get A, second bump to T, but then a huge increase in memory to get to S. Also, the time increases substantially (notice the extra time in between INTERVAL 2 and INTERVAL 3).

![Image](https://github.com/user-attachments/assets/c9e322d1-66b9-4000-9357-45d8eecf5e47)

![Image](https://github.com/user-attachments/assets/f9acebba-0f5b-4ee2-9519-5d0ad6496727)

### Versions

all versions

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @amjames @bhosmer @jcaip

**Comment by aartbik** [comment]
Notifying my sparse friends @ezyang @pearu

**Comment by aartbik** [comment]
FWIW, the three "spikes" in the CSR plot are caused by three  at::native::arange() operations

first two spikes in _not_zero_mask_to_col_row_indices() for the rows and columns
third spike in _mask_to_indices()

**Comment by aartbik** [comment]
Just hacking around a bit, when I replace the third spiking code
```
 return = at::native::arange(
      mask.numel(), at::kLong, kStrided, mask.device())
      .masked_select(mask);

```
with just

` return at::flatten( at::nonzero(mask) );`

I completely eliminate the third spike, and still get the same result. Is this on the right track?

![Image](https://github.com/user-attachments/assets/d7c508fc-37c2-4de9-b20d-bfe9b73fa847)

**Comment by aartbik** [comment]
I believe I have a fix for the first two spikes as well. The to_sparse() and to_sparse_csr() now behave the same!

![Image](https://github.com/user-attachments/assets/2b95f8bd-110e-44c0-a6b3-a68109c085ca)

**Comment by pearu** [comment]
From reading the `_not_zero_mask_to_col_row_indices` code with an observation that `arange` is memory greedy, there exists a way to optimize this code by using single arange result for both `col_indices` and `row_indices` (we have used this optimization in other places, IIRC).
Also, since conversion of dense to CSR is a special case of the conversion of dense to COO, I would expect that their memory consumption characteristics could be made comparable.

^ @amjames

**Comment by aartbik** [comment]
Yes, the second fix I have in mind as as follows:

```
static std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
    Tensor not_zero_mask,
    ScalarType index_dtype,
    Device index_device) {
 Tensor nz = not_zero_mask.nonzero().transpose(0, 1);
  return std::pair<Tensor, Tensor>(nz[1], nz[0]);
}
```

This, together with the fix already out of review gets almost exactly the same behavior for COO and CSR.

**Comment by aartbik** [comment]
I did a write-up in https://dev-discuss.pytorch.org/t/performance-of-activation-sparsity/2915
